R/sim_sce.R

```diff
@@ -0,0 +1,58 @@
+#' Create a random SingleCellExperiment object
+#'
+#' @param n_genes The number of genes to simulate
+#' @param n_cells The number of cells to create.
+#' @param n_empty How many "empty" droplets to create.
+#'   This will generally be much larger than the number of cells.
+#' @param n_groups How many expression groups (cell clusters, roughly) to create.
+#'
+#' @return A SingleCellExperiment object.
+#'
+#' @importFrom stats rexp rpois
+#' @importFrom utils head
+#'
+#' @examples
+#' \dontrun{
+#'   sim_sce(n_genes = 100, n_cells = 100, n_empty = 2000, n_groups = 4)
+#' }
+#'
+sim_sce <- function(n_genes = 200, n_cells = 100, n_empty = 1000, n_groups = 3){
+  if (n_genes < 1){
+    stop("n_genes must be a positive number.")
+  }
+  if (n_cells < 1){
+    stop("n_cells must be a postive number.")
+  }
+  # enforce some minimums
+  n_empty <- max(n_empty, 0)
+  n_groups <- max(n_groups, 1)
+
+  if (n_cells + n_empty > 1000000){
+    warning("Did you really want to simulate more than a million droplets?")
+  }
+
+  # generate a random distribution of gene expression for each group
+  gene_expr <- rexp(n_genes * n_groups, 0.2)
+  # expression for each gene x cell
+  cell_expr <- matrix(rpois(n_genes * n_cells, lambda = gene_expr), nrow = n_genes)
+
+  # empty drops will have expression calculated from the overall mean
+  empty_expr <- rowMeans(cell_expr) * 0.01
```

Comment:
This is just so I understand, but why can't we just make a vector of zeros here?

Reply:
I wanted it to have a whiff of realism? The idea is that there are droplets with a few reads from the soup, so the expression counts will be based on the average expression of the "real" cells, just much lower. I actually checked whether EmptyDrops would toss these cells, and it does... probably mostly because they have lower expression, but it keeps the "real" cells, so that seemed good enough for now.

Reply:
Ah that makes a lot of sense. I had forgotten that empty drops actually do have some reads in them sometimes.